src/coder/graph.py

```diff
@@ -1,37 +1,110 @@
+    """Configuration for a coder instance."""
+
+    name: str
+    system_prompt: str
+    github_tools: List[str]
+
+    def graph_builder(self, github_toolset: list[Tool]):
+        builder = graph_builder(self.filter_tools(github_toolset), self.system_prompt)
+        builder.name = self.name
+        return builder
+
+    def filter_tools(self, tools: List[Tool]) -> List[Tool]:
+        """Filter tools to only include those specified in github_tools.
+
+        Args:
+            tools: List of all available tools
+
+        Returns:
+            List of tools that match the names in github_tools
+
+        Raises:
+            AssertionError: If the number of filtered tools doesn't match github_tools
+        """
+        filtered_tools = [tool for tool in tools if tool.name in self.github_tools]
+        assert len(filtered_tools) == len(self.github_tools), (
+            f"Tool mismatch. Expected {len(self.github_tools)} tools, got {len(filtered_tools)}. "
+            f"Expected tools: {self.github_tools}, Got tools: {[t.name for t in filtered_tools]}"
+        )
+        return filtered_tools
+
+
+def coder_new_pr_config():
+    """Instance config for Coder | New PR. Invoked when directed to create a new PR."""
+    return CodeInstanceConfig(
+        name="Coder | New PR",
+        system_prompt=NEW_PR_SYSTEM_PROMPT,
+        github_tools=[
+            "set_active_branch",
+            "create_a_new_branch",
+            "get_files_from_a_directory",
+            "create_pull_request",
+            "create_file",
+            "update_file",
+            "read_file",
+            "delete_file",
+        ],
+    )
+
+
+def coder_change_request_config():
+    """Instance config for Coder | Change Request. Invoked when directed to make a changes to an existing PR."""
+    return CodeInstanceConfig(
+        name="Coder | Change Request",
+        system_prompt=CHANGE_REQUEST_SYSTEM_PROMPT,
+        github_tools=[
+            "set_active_branch",
+            "get_files_from_a_directory",
+            "create_file",
+            "update_file",
+            "read_file",
+            "delete_file",
+            "get_pull_request",
+            "list_pull_requests_files",
+        ],
+    )
+
+
 class CallModel:
-    def __init__(self, github_tools: list[Tool]):
+    def __init__(self, github_tools: list[Tool], system_prompt: str):
         self.github_tools = github_tools
+        self.system_prompt = system_prompt
 
     async def __call__(self, state: State) -> dict:
-        system_msg = SystemMessage(content=SYSTEM_PROMPT)
+        system_msg = SystemMessage(content=self.system_prompt)
         messages = [system_msg] + state["messages"]
         messages_after_invoke = await llm.bind_tools(self.github_tools).ainvoke(
             messages
         )
         return {"messages": messages_after_invoke}
 
 
-def graph_builder(github_toolset: list[Tool]) -> StateGraph:
+def graph_builder(github_toolset: list[Tool], system_prompt: str):
```

Comment:
should this be `_graph_builder` now that it's not exported? (Not a python expert, so school me if not...)